pystac/summaries.py

```diff
@@ -0,0 +1,270 @@
+import sys
+import json
+import numbers
+import urllib.request
+from enum import Enum
+
+import pystac
+from pystac.utils import get_required
+
+from typing import (
+    Any,
+    Dict,
+    Generic,
+    List,
+    Optional,
+    Union,
+    cast,
+    TypeVar,
+    Iterable,
+    TYPE_CHECKING,
+)
+
+if sys.version_info >= (3, 8):
+    from typing import Protocol
+else:
+    from typing_extensions import Protocol
+
+if TYPE_CHECKING:
+    from pystac.item import Item as Item_Type
+    from pystac.collection import Collection as Collection_Type
+
+from abc import abstractmethod
+
+
+class Comparable(Protocol):
+    """Protocol for annotating comparable types."""
+
+    @abstractmethod
+    def __lt__(self: "T", x: "T") -> bool:
+        pass
+
+
+T = TypeVar("T", bound=Comparable)
+
+
+class RangeSummary(Generic[T]):
+    def __init__(self, minimum: T, maximum: T):
+        self.minimum = minimum
+        self.maximum = maximum
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {"minimum": self.minimum, "maximum": self.maximum}
+
+    def update_with_value(self, v: T) -> None:
+        self.minimum = min(self.minimum, v)
+        self.maximum = max(self.maximum, v)
+
+    @classmethod
+    def from_dict(cls, d: Dict[str, Any]) -> "RangeSummary[T]":
+        minimum: T = get_required(d.get("minimum"), "RangeSummary", "minimum")
+        maximum: T = get_required(d.get("maximum"), "RangeSummary", "maximum")
+        return cls(minimum=minimum, maximum=maximum)
+
+
+FIELDS_JSON_URL = "https://cdn.jsdelivr.net/npm/@radiantearth/stac-fields/fields.json"
```

Comment:
```suggestion
FIELDS_JSON_URL = "https://cdn.jsdelivr.net/npm/@radiantearth/stac-fields/fields-normalized.json"
```

This must be the normalized version, otherwise, you'll run into bugs here as some fields may not be resolved yet.